### DISCUSSION TITLE
Blacklist and new directories

### DISCUSSION THREAD

**Discussion by jfernandez50** [netblue30/firejail#5037] [Q&A] [answered]
Hi,

If a profile lists a non-existing directory as blacklist, a process is running with that profile, and another process creates that directory, the first process will be able to list the directory.

For example, terminal 1:
```bash
user@host:/tmp/fjtest$ cat profile 
blacklist foo
blacklist bar
user@host:/tmp/fjtest$ ls
foo profile
user@host:/tmp/fjtest$ ls foo
foo.txt
user@host:/tmp/fjtest$ firejail --profile=profile
Reading profile profile
Parent pid 1468983, child pid 1468984
Child process initialized in 14.94 ms
user@host:/tmp/fjtest$ ls
foo profile
user@host:/tmp/fjtest$ ls foo
ls: cannot open directory 'foo': Permission denied
user@host:/tmp/fjtest$
```

Without exiting terminal 1, in terminal 2 (outside the sandbox):
```bash
user@host:/tmp/fjtest$ mkdir bar
user@host:/tmp/fjtest$ echo bar > bar/bar.txt
```

Back in terminal 1 (still running and in the sandbox):
```bash
user@host:/tmp/fjtest$ ls 
bar foo profile
user@host:/tmp/fjtest$ ls bar
bar.txt
user@host:/tmp/fjtest$ cat bar/bar.txt
bar
user@host:/tmp/fjtest$
```

Even if bar is explicitly mentioned in the profile. If I stop the process in terminal 1 and restart it, then it gets fixed:
```bash
user@host:/tmp/fjtest$ exit

Parent is shutting down, bye...
user@host:/tmp/fjtest$ firejail --profile=profile
Reading profile profile
Parent pid 1469460, child pid 1469461
Child process initialized in 23.20 ms
user@host:/tmp/fjtest$ ls foo
ls: cannot open directory 'foo': Permission denied
user@host:/tmp/fjtest$ ls bar
ls: cannot open directory 'bar': Permission denied
user@host:/tmp/fjtest$
```

Is there any way to establish in the profile directories that do not exist yet but that we know that they are going to exist? 

Or, if we know when the directories are created, is there any way to trigger an event to firejail to force the sandbox to reload the rules based described on the profile file; or even to add new rules?

Just in case, my real use case is that we have multiple USB devices connected of one type of device to a computer, and a third party closed-source program that uses them, but can only handle one at a time, and tends to fail when there are multiple of them. I am using firejail to run the program multiple times isolated, so each process only sees one device of `/sys/bus/usb` and not the others. When everything is connected, it works well. But whenever a new device is added, the existing processes will see the new device even if it was blacklisted in their profiles, and therefore they start to fail.

Thank you!

**Comment by rusty-snake** [comment, marked as answer]
> force the sandbox to reload the rules based described on the profile file

You can not "reload", you can only "restart" the sandbox.

> Or, if we know when the directories are created, is there any way to trigger an event to firejail to force the sandbox to reload the rules based described on the profile file; or even to add new rules?

This would always be racey, it wouldn't?

> /sys/bus/usb

If it would be something else I'd suggested to use `whitelist`.

**Reply by jfernandez50** [reply]
Thank you for your answer.

Just to clarify: "restart the sandbox" implies restarting the process, right?

Therefore I assume this is not possible to block a folder (e.g., `blacklist /path1`) if that folder does not exist in the start of the process (other than with `whitelist` that doesn't work in `/sys/`).

Thank you!

**Reply by rusty-snake** [reply]
> Just to clarify: "restart the sandbox" implies restarting the process, right?

Yes, like close program and reopen it.

**Reply by jfernandez50** [reply]
Thank you!

**Comment by ghost** [comment]
> But whenever a new device is added, the existing processes will see the new device even if it was blacklisted in their profiles, and therefore they start to fail.

As pointed out by @rusty-snake your best Firejail option - whitelist - doesn't support /sys. So you need something else in tandem with Firejail for this specific use case IMO. I'd look into `udev rules` to lock down newly added devices until the third-party program can handle them (if that kind of sequential processing is desirable/possible). Or something like `usbguard` to the same effect.

**Reply by jfernandez50** [reply]
Thank you for your answer.

No, unfortunately, I do need to support all the USB devices at the same time, each one in a different process; and they are used concurrently by different users; so when I unplug and plug one device, it can't affect the others / restart it. I guess that therefore `usbguard` wouldn't work in this context. `udev rules` might if I create a user per device (quite desperate, but I might have to do it, they're only 16), but not otherwise as I do not see another way to apply a policy by program or a way to change the serial identifier (the third-party program can filter by serial identifier; but all the devices have the same serial identifier).

Before with firejail it worked perfectly fine with `blacklist`, but only if we plugged everything in the beginning and never unplug and replug (otherwise, we need to restart everything and affect users).

I'll trying to solve it with apparmor, using deny for certain /sys/ paths.

Thank you for your feedback!!!
